csharp-dotnet/README.md

```diff
@@ -0,0 +1,110 @@
+- A semicolon terminates a statement, and goes on the same line as the end of the statement.
+- Declare member variables at the top of a class, with static variables at the very top.
+
+**<u>Always Use Access Modifiers</u>**
+
+Use the lowest necessary modifier - principle of least privilege.
+
+**<u>Use Auto Properties</u>**
+
+Unless you need control over the backing field, don't create one. Use the auto property.
+
+**<u>Code File Organization</u>**
+
+Use a single file for each class, interface, struct or enum.
+
+
+
+## Naming Conventions
+
+**<u>Pascal Casing</u>**
+
+The first letter in the identifier and the first letter of each subsequent concatenated word are capitalized. Use Pascal casing for class names, file names, namespace names, method names and public member names.
+
+`FreezingPoint	`
+
+**<u>Camel Casing</u>**
+
+The first letter in the identifier is lowercase and the first letter of each subsequent concatenated word is capitalized. Use Camel casing for local variables and method parameters that are not publicly available
+
+`freezingPoint`
+
+**<u>Snake Casing</u>**
+
+The first letter of each word is lowercase. Each space in the identifier is replaced by an underscore (_) character. This convention may be used on descriptive names for unit test classes.
+
+`snake_case_used_for_test_classes`
+
+**<u>Upper Snake Casing</u>**
+
+All letters in the identifier are capitalized (aka upper snake case). This is only used for constants. For example:
+
+`FREEZING_POINT`
+
+**<u>Do's and Don'ts</u>** 
+
+- Do not use Hungarian notation to add type identification to identifiers (e.g., string strName).
+- Do not use underscores in identifiers. *Exceptions: unit test class names, and private static variables* *(e.g., _injectedService).*
+- Prefix interfaces with the letter capital letter I.
+- Name source files according to their main classes.
+- Use singular names for enums. *Exception: bit field enums.*
+
+
+
+## Commenting Conventions
+
+- Place the comment on a separate line, not at the end of a line of code.
+
+- Begin comment text with an uppercase letter.
+
+- End comment text with a period.
+
+- Insert one space between the comment delimiter (//) and the comment text, as shown in the following example.
+
+  ```csharp
+  // The fox jumped over the dog.
+  ```
+
+- Do not create formatted blocks of asterisks around comments.
+
+  
+
+## Language Guidelines
+
+**<u>String Data Types</u>**
+
+- Use [string interpolation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) to concatenate short strings.
+- To append strings in loops, especially with large amounts of text, use a StringBuilder object.
+
+<u>**Implicitly Typed Local Variables**</u>
+
+- Use implicit typing (i.e., var) when the type of the variable is obvious from the right side of the assignment, or when precise type is not important.
+- Do not use var when the type is not apparent from the right side of the assignment.
```

Comment:
I think this is a pretty important point. Do you happen to have a couple examples or guidelines to help determine what's obvious/important vs what wouldn't be obvious/important? Just to help reduce any subjectiveness about when `var` is ok to use.

Reply:
Implicitly Typed Local Variables
Use implicit typing for local variables when the type of the variable is obvious from the right side of the assignment, or when the precise type is not important.

C#

Copy
// When the type of a variable is clear from the context, use var
// in the declaration.
var var1 = "This is clearly a string.";
var var2 = 27;
Do not use var when the type is not apparent from the right side of the assignment.

C#

Copy
// When the type of a variable is not clear from the context, use an
// explicit type. You generally don't assume the type clear from a method name.
// A variable type is considered clear if it's a new operator or an explicit cast.
int var3 = Convert.ToInt32(Console.ReadLine());
int var4 = ExampleClass.ResultSoFar();
Do not rely on the variable name to specify the type of the variable. It might not be correct.

C#

Copy
// Naming the following variable inputInt is misleading.
// It is a string.
var inputInt = Console.ReadLine();
Console.WriteLine(inputInt);
Avoid the use of var in place of dynamic.

Use implicit typing to determine the type of the loop variable in for loops.

The following example uses implicit typing in a for statement.

C#

Copy
var phrase = "lalalalalalalalalalalalalalalalalalalalalalalalalalalalalala";
var manyPhrases = new StringBuilder();
for (var i = 0; i < 10000; i++)
{
    manyPhrases.Append(phrase);
}
//Console.WriteLine("tra" + manyPhrases);
Do not use implicit typing to determine the type of the loop variable in foreach loops.

The following example uses explicit typing in a foreach statement.

C#

Copy
foreach (char ch in laugh)
{
    if (ch == 'h')
        Console.Write("H");
    else
        Console.Write(ch);
}
Console.WriteLine();

Reply:
Great examples Troy.

What are your thoughts about string.format when you have 1 to 3 variables
Example:
var firstInput = "error message 1";
var secondInput = "Some output";

Using String Builder:
var msg = new StringBuilder("The following error occurred: ");
msg.Append(firstInput);
msg.Append(". Error Output: ");
msg.Append(secondInput);

String Format
var msg = string.Format("The following error occurred: {0}. Error Output: {1}", firstInput, secondInput);

Reply:
I debated whether to even include the implicit typing section, since it's always going to be subjective and the compiler can't enforce. The Microsoft standards expand on it, so I'll try to condense that with Troy and Pat's comments.

Reply:
If you have C# 6 or better, [string interpolation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) is recommended. It's easier to write and to read.

var msg = $"The following error occurred: {firstInput}. Error Output: {secondInput}";